Port to dune & various bug fixes

let-def opened the pull request:
This main change in this PR is to switch the build system to dune.

This is an error-prone process: I might have forgotten some targets and it will change your programming habit a bit (for instance the documentation output directory changed). I tried to stay as close as possible to the original settings. I haven't touched the `GNUMakefile` recipes though.

Beside that the following things have changed:
- in the blog post, the `flow` example was changed to take a document (`flow 1 [...]` to `flow (break 1) [...]`)
- minor warnings have been adressed
- the renderer to `Format.formatter` has changed a bit.

I believe the previous implementation was wrong because it used a mix of `pp_print_char` and direct access to formatter output functions:
- `pp_print_char` delays the output as it accumulates formatting instructions and waits for enough context to be known for a decision to be taken
- directly accessing output functions is eager, so contents output using the substring interface would appear before the char one; it also bypasses any content and formatting instructions the formatter might had, leading to confusing behaviors.

The new implementation is slightly less efficient because new strings have to be allocated (it is not possible to send a sub-string to a formatter), but it should integrate well with other use of `Format`, such as the toplevel printers.
For instance, `#install_printer PPrint.ToFormatter.compact;;` now has a reasonable behavior for printing documents.

# String or text

I choose to call `Format.pp_print_string` but `Format.pp_print_text` is another reasonable option, depending on the behavior we want when outputting to a formatter.
Actually, I think calling `pp_print_text` is even better (see https://caml.inria.fr/pub/docs/manual-ocaml/libref/Format.html#VALpp_print_text) as it will try to honor the local Formatter indentation and breaking rules.

fpottier commented on pprint.opam:
```diff
@@ -8,13 +8,12 @@ authors: [
   "ocaml" {>= "4.02"}
-  "ocamlfind" {build}
-  "ocamlbuild" {build}
+  "dune" { "1.11" }
```
Should this be `>= "1.11"`?

let-def replied:
Thanks, I fixed that and put `>= "1.3"` as I tested and it worked well with this old version of dune. This should not put unnecessary pressure on people to upgrade their packages.

fpottier commented:
Why is the file `pprint.install` under version control? Should it not be ignored, as it is generated by `dune`?

fpottier commented:
Regarding the choice between `Format.pp_print_string` and `Format.pp_print_text`, I don't have a strong opinion, as I do not know `Format` very well at all. If you think the latter is preferable, that's fine with me.

let-def commented:
I removed the `pprint.install` file and switched the formatter renderer to use `pp_print_text`.
Here is a sample toplevel session:
```
# #require "pprint";;
# #install_printer PPrint.ToFormatter.compact;;
# #use "src/PPrintTest.ml";;
val paragraph : string -> PPrintEngine.document = <fun>
val document : PPrint.document = TITLE: PPrint
  AUTHORS: François Pottier and Nicolas Pouillard
  ABSTRACT: This is an adaptation of Daan Leijen's "PPrint" library, which
  itself is based on the ideas developed by Philip Wadler in "A Prettier
  Printer". For more information about Wadler's and Leijen's work, please
  consult the following references:  http://www.cs.uu.nl/~daan/pprint.html
  http://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf  To
  install PPrint, type "make -C src install". ocamlfind is required.  The
  documentation for PPrint is built by "make doc" and is found in the file
  doc/index.html.
...
```

fpottier commented on src/PPrintEngine.ml:
```diff
@@ -79,14 +79,17 @@ class buffer_output buffer = object
+
   method substring str ofs len =
-    let str' =
+    Format.pp_print_string fmt (
```
Is it intentional that you still use `pp_print_string` whereas you said you would use `pp_print_text`?

fpottier replied:
(awaiting your answer on this point?)

let-def replied:
That was a mistake, sorry for that. It is now fixed, thanks :).

trefis commented on Makefile:
```diff
@@ -1,4 +1,18 @@
-all install uninstall reinstall clean doc test:
-	$(MAKE) -C src $@
+all:
+	dune build -p pprint
```
What's the reason for this `-p pprint`?

let-def replied:
It was just the same as in the opam file. Though @trefis just told me that this will disable warnings during development so I will remove `-p pprint` :)
